### PR TITLE
[stable/postgresql] PostgreSQL maintenance proposal

### LINKED ISSUE #7956: [stable/postgresql] PostgreSQL maintenance proposal
I just checked PostgreSQL chart and I have found some aspects that could be improved.  I tried to contact maintainers but according to [stable `Chart.yaml`](https://github.com/helm/charts/blob/master/stable/postgresql/Chart.yaml#L15), there are no official maintainers at this time. I've been doing a comparison between the stable and the Bitnami chart, below you can see the results.

**Differences between [upstream PostgreSQL chart](https://github.com/helm/charts/tree/master/stable/postgresql) and the [Bitnami one](https://github.com/bitnami/charts/tree/master/bitnami/postgresql/):**
- Currently, stable version is `9.6.2` from 6 months ago while Bitnami version is `10.5.0` from 1 month ago.
In fact, the stable chart is not up-to-date in the `9.6` branch because the current minor is `9.6.10` according to https://www.postgresql.org/support/versioning/.
Bitnami has different systems that check upstream versions and updates the images following an automated test + release process.
- Bitnami one has replication (disabled by default), the stable one doesn't include this feature.
- Bitnami one has _Pod Security Context_, the stable one doesn't include this feature.
- Bitnami one has `values.yaml` and `values-production.yaml`, the stable only `values.yaml`.
- Upstream one has _NetworkPolicy_, the Bitnami one doesn't include this feature (but we can add it).
- Both use metrics _exporter_, _secrets_ and _PVC_.

Due to the lack of maintainers and updates (apart of some features), from Bitnami we are willing to help maintain this chart, we can adopt it and include it in our workflow to keep it updated as we do with the rest of the charts.

What do you think @prydonius @mattfarina @desaintmartin

### PR DESCRIPTION
**Pull request by carrodher**
#### What this PR does / why we need it:
This PR is intended in case the PostgreSQL maintenance proposal (#7956) is accepted. 
In this way, you can see some differences between both charts, as well as work on applying new features or improve existing ones.

**Differences between [upstream PostgreSQL chart](https://github.com/helm/charts/tree/master/stable/postgresql) and the [Bitnami one](https://github.com/bitnami/charts/tree/master/bitnami/postgresql/):**
- Currently, stable version is `9.6.2` from 6 months ago while Bitnami version is `10.5.0` from 1 month ago.
In fact, the stable chart is not up-to-date in the `9.6` branch because the current minor is `9.6.10` according to https://www.postgresql.org/support/versioning/.
Bitnami has different systems that check upstream versions and updates the images following an automated test + release process.
- Bitnami one has replication (disabled by default), the stable one doesn't include this feature.
- Bitnami one has _Pod Security Context_, the stable one doesn't include this feature.
- Bitnami one has `values.yaml` and `values-production.yaml`, the stable only `values.yaml`.
- Upstream one has _NetworkPolicy_, the Bitnami one doesn't include this feature (but we can add it).
- Both use metrics _exporter_, _secrets_ and _PVC_.

#### Which issue this PR fixes
- fixes #7956

#### Checklist
- [X] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [X] Chart Version bumped
- [X] Variables are documented in the README.md

### CODE REVIEW

File: stable/postgresql/Chart.yaml [hunk 1]
```diff
@@ -1,16 +1,19 @@
 name: postgresql
-version: 0.18.1
-appVersion: 9.6.2
-description: Object-relational database management system (ORDBMS) with an emphasis on extensibility and on standards-compliance.
+version: 1.0.0
```
Comment (scottrigby) [inline stable/postgresql/Chart.yaml]: What are the steps to upgrade to this version? Per https://github.com/helm/charts/issues/5657 if there are any steps please document them in an "Upgrade" section of the README.

File: stable/postgresql/Chart.yaml [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 home: https://www.postgresql.org/
-icon: https://www.postgresql.org/media/img/about/press/elephant.png
+icon: https://bitnami.com/assets/stacks/postgresql/img/postgresql-stack-110x117.png
 sources:
-- https://github.com/kubernetes/charts
-- https://github.com/docker-library/postgres
-maintainers: []
+- https://github.com/bitnami/bitnami-docker-postgresql
+maintainers:
+- name: Bitnami
```
Comment (scottrigby) [inline stable/postgresql/Chart.yaml:17]: Per charts chat Today, please also add appropriate GitHub users to a new OWNERS file.

Comment (carrodher) [reply stable/postgresql/Chart.yaml:17]: Thanks for point it. I'll copy the `OWNERS` from other Bitnami maintained chart

File: stable/postgresql/README.md [hunk 3]
```diff
@@ -33,123 +33,132 @@ The command deploys PostgreSQL on the Kubernetes cluster in the default configur
...
+|----------------------------|-------------------------------------------|---------------------------------------------------------- |
+| `image.registry`           | PostgreSQL image registry                 | `docker.io`                                               |
+| `image.repository`         | PostgreSQL Image name                     | `bitnami/postgresql`                                      |
+| `image.tag`                | PostgreSQL Image tag                      | `{VERSION}`                                               |
+| `image.pullPolicy`         | PostgreSQL image pull policy              | `Always`                                                  |
+| `image.pullSecrets`        | Specify image pull secrets                | `nil` (does not add image pull secrets to deployed pods)  |
+| `image.debug`              | Specify if debug values should be set     | `false`                                                   |
+| `replication.enabled`      | Would you like to enable replication      | `false`                                                   |
```
Comment (juan131) [inline stable/postgresql/README.md:56]: I'd enabled replication by default as we do on MariaDB.

File: stable/postgresql/values.yaml [hunk 4]
```diff
@@ -1,122 +1,91 @@
...
+  enabled: true
+  fsGroup: 1001
+  runAsUser: 1001
 
-## Create a database user
-## Default: postgres
-# postgresUser:
-## Default: random 10 character string
-# postgresPassword:
+replication:
```
Comment (juan131) [inline stable/postgresql/values.yaml:34]: Shouldn't we enabled replication by default?

File: stable/postgresql/Chart.yaml (hunk 1 above)
Comment (juan131) [reply stable/postgresql/Chart.yaml]: +1

File: stable/postgresql/values.yaml (hunk 4 above)
Comment (carrodher) [reply stable/postgresql/values.yaml:34]: We are enabling it by default in `values-production.yaml`

File: stable/postgresql/README.md (hunk 3 above)
Comment (carrodher) [reply stable/postgresql/README.md:56]: We are enabling it by default in `values-production.yaml`

File: stable/postgresql/Chart.yaml (hunk 1 above)
Comment (javsalgar) [reply stable/postgresql/Chart.yaml]: +1

File: stable/postgresql/templates/NOTES.txt [hunk 5]
```diff
@@ -1,41 +1,53 @@
...
-{{ template "postgresql.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
+{{- if contains .Values.service.type "LoadBalancer" }}
+{{- if not .Values.postgresqlPassword }}
+-------------------------------------------------------------------------------
+ WARNING
 
-{{- if .Values.existingSecret }}
-If you have not already created the postgres admin secret:
+    By specifying "serviceType=LoadBalancer" and not specifying "postgresqlPassword"
+    you have most  likely exposed the PostgreSQL service externally without any
```
Comment (tompizmor) [inline stable/postgresql/templates/NOTES.txt]: extra space in `most likely`

File: stable/postgresql/templates/NOTES.txt [hunk 6]
```diff
@@ -1,41 +1,53 @@
...
+    authentication mechanism.
 
-   kubectl create secret generic {{ .Values.existingSecret }} --namespace {{ .Release.Namespace }} --from-file=./postgres-password
-{{ else }}
-To get your user password run:
+    For security reasons, we strongly suggest that you switch to "ClusterIP" or
+    "NodePort". As alternative, you can also specify a valid password on the
```
Comment (tompizmor) [inline stable/postgresql/templates/NOTES.txt]: `As alternative` -> `As an alternative`

File: stable/postgresql/templates/configmap.yaml [hunk 7]
```diff
@@ -1,16 +1,13 @@
+{{ if  (.Files.Glob "files/postgresql.conf") }}
 apiVersion: v1
 kind: ConfigMap
 metadata:
-  name: {{ template "postgresql.fullname" . }}
+  name: {{ template "postgresql.fullname" . }}-configuration
   labels:
-    app: {{ template "postgresql.name" . }}
+    app: "{{ template "postgresql.name" . }}"
```
Comment (tompizmor) [inline stable/postgresql/templates/configmap.yaml]: delete the quotes

File: stable/postgresql/templates/_helpers.tpl [hunk 8]
```diff
@@ -3,26 +3,30 @@
 Expand the name of the chart.
 */}}
 {{- define "postgresql.name" -}}
-{{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" -}}
+{{- default .Chart.Name .Values.nameOverride | trunc 24 -}}
```
Comment (tompizmor) [inline stable/postgresql/templates/_helpers.tpl]: leave it as it was in the previous version. trunc at 63 characters and add the trimSuffix

```
{{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" -}}
```

File: stable/postgresql/templates/_helpers.tpl [hunk 9]
```diff
@@ -3,26 +3,30 @@
...
 {{- end -}}
 
 {{/*
 Create a default fully qualified app name.
-We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
-If release name contains chart name it will be used as a full name.
+We truncate at 24 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
```
Comment (tompizmor) [inline stable/postgresql/templates/_helpers.tpl]: we should truncate at 63 characters

File: stable/postgresql/templates/initialization-configmap.yaml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "postgresql.fullname" . }}-init-scripts
+  labels:
+    app: "{{ template "postgresql.name" . }}"
```
Comment (tompizmor) [inline stable/postgresql/templates/initialization-configmap.yaml]: delete the quotes

File: stable/postgresql/templates/metrics-deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,75 @@
+{{- if .Values.metrics.enabled }}
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "postgresql.fullname" . }}-metrics
+  labels:
+    app: {{ template "postgresql.name" . }}
+    chart: {{ template "postgresql.chart" . }}
+    release: "{{ .Release.Name }}"
```
Comment (tompizmor) [inline stable/postgresql/templates/metrics-deployment.yaml]: `{{ .Release.Name | quote }}` the same for other labels in this file and others

File: stable/postgresql/templates/metrics-svc.yaml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  labels:
+    app: {{ template "postgresql.name" . }}
+    chart: {{ template "postgresql.chart" . }}
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+  annotations:
+{{ toYaml .Values.metrics.service.annotations | indent 4 }}
+spec:
+  type: {{ .Values.metrics.service.type }}
+  {{ if eq .Values.metrics.service.type "LoadBalancer" -}} {{ if .Values.metrics.service.loadBalancerIP -}}
```
Comment (tompizmor) [inline stable/postgresql/templates/metrics-svc.yaml]: I think you can combine this `if` in a single line

File: stable/postgresql/templates/statefulset-slaves.yaml [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        ports:
+        - name: postgresql
+          containerPort: {{ .Values.service.port }}
+        {{- if .Values.livenessProbe.enabled }}
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - exec pg_isready -U {{ default "" .Values.postgresqlUsername | quote }} --host $POD_IP
```
Comment (tompizmor) [inline stable/postgresql/templates/statefulset-slaves.yaml]: delete the `default`

File: stable/postgresql/templates/statefulset-slaves.yaml [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+          successThreshold: {{ .Values.livenessProbe.successThreshold }}
+          failureThreshold: {{ .Values.livenessProbe.failureThreshold }}
+        {{- end }}
+        {{- if .Values.readinessProbe.enabled }}
+        readinessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - exec pg_isready -U {{ default "" .Values.postgresqlUsername | quote }} --host $POD_IP
```
Comment (tompizmor) [inline stable/postgresql/templates/statefulset-slaves.yaml]: delete the default value

File: stable/postgresql/templates/statefulset.yaml [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        ports:
+        - name: postgresql
+          containerPort: {{ .Values.service.port }}
+        {{- if .Values.livenessProbe.enabled }}
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - exec pg_isready -U {{ default "" .Values.postgresqlUsername | quote }} --host $POD_IP
```
Comment (tompizmor) [inline stable/postgresql/templates/statefulset.yaml]: delete the default value

File: stable/postgresql/templates/statefulset.yaml [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+          successThreshold: {{ .Values.livenessProbe.successThreshold }}
+          failureThreshold: {{ .Values.livenessProbe.failureThreshold }}
+        {{- end }}
+        {{- if .Values.readinessProbe.enabled }}
+        readinessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - exec pg_isready -U {{ default "" .Values.postgresqlUsername | quote }} --host $POD_IP
```
Comment (tompizmor) [inline stable/postgresql/templates/statefulset.yaml]: delete the default value

File: stable/postgresql/templates/svc-headless.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+spec:
+  type: ClusterIP
+  clusterIP: None
+  ports:
+  - name: postgresql
+    port: 5432
+    targetPort: postgresql
+  selector:
+    app: {{ template "postgresql.name" . }}
+    release: "{{ .Release.Name }}"
```
Comment (tompizmor) [inline stable/postgresql/templates/svc-headless.yaml]: use `{{ .Release.Name | quote }}`

**Review by tompizmor — CHANGES_REQUESTED** [review]

File: stable/postgresql/templates/NOTES.txt [hunk 18]
```diff
@@ -1,41 +1,53 @@
...
+    "postgresqlPassword" parameter.
```
Comment (javsalgar) [inline stable/postgresql/templates/NOTES.txt:13]: We should add something about how to connect when the networkpolicy is enabled (it was in the previous notes). Not sure if I missed it

File: stable/postgresql/templates/statefulset-slaves.yaml [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+      containers:
+      - name: "{{ template "postgresql.fullname" . }}"
+        image: "{{ template "postgresql.image" . }}"
+        imagePullPolicy: "{{ .Values.image.pullPolicy }}"
+        resources:
+{{ toYaml .Values.Resources | indent 10 }}
+        env:
+        {{- if .Values.image.debug}}
+        - name: BASH_DEBUG
+          value: 1
```
Comment (javsalgar) [inline stable/postgresql/templates/statefulset-slaves.yaml:59]: Please test this. I think it may not work

**Review by javsalgar — CHANGES_REQUESTED** [review]
Small changes, the rest LGTM

**Comment by carrodher** [comment]
Hi all @scottrigby @prydonius @tompizmor @juan131 @javsalgar 

I just implemented the requested changes (and merged latest features), can you take a look? Thanks!

File: stable/postgresql/templates/configmap.yaml [hunk 20]
```diff
@@ -1,16 +1,13 @@
...
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
 data:
-  {{- if .Values.metrics.customMetrics }}
-  custom-metrics.yaml: {{ toYaml .Values.metrics.customMetrics | quote }}
-  {{- end }}
-  {{- if .Values.pgHbaConf }}
-  pg_hba.conf: {{ .Values.pgHbaConf | quote }}
```
Comment (desaintmartin) [inline stable/postgresql/templates/configmap.yaml:15]: Do you intend to keep support somehow of pg_hba.conf?

**Review by juan131 — APPROVED** [review]
LGTM

File: stable/postgresql/templates/configmap.yaml (hunk 20 above)
Comment (carrodher) [reply stable/postgresql/templates/configmap.yaml:15]: Hi! At this moment the chart supports custom configuration using `postgresql.conf`, but I am working to provide support for both (`postgresql.conf` and `pg_hba.conf`). But this change not only requires modifications in the chart (this is the easy part), it also requires changes in the Docker image.

TL;DR I am working to modify the Docker image to support it.

Do you think it is a blocking issue to merge it or can we merge the PR and send a new one with the `ph_hba.conf` changes after some days?

Comment (desaintmartin) [reply stable/postgresql/templates/configmap.yaml:15]: It is blocking for me (I have a dozen of postgres running using that feature) for the long term but I guess it can wait for a few days with no harm and in case of problem I can use my own fork.
Tell me if I can help.

Comment (carrodher) [reply stable/postgresql/templates/configmap.yaml:15]: Thanks!
I just added some logic to support it at chart level 92498ba, feel free to review it!
In the other hand I continue working on the Docker image stuff

File: stable/postgresql/files/README.md [hunk 21]
```diff
@@ -0,0 +1 @@
+Copy here your postgresql.conf file to use it as a config map.
```
Comment (tompizmor) [inline stable/postgresql/files/README.md]: Update this message as mounting the `pg_hba.conf` will also be possible

File: stable/postgresql/templates/metrics-deployment.yaml [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+      {{- if .Values.metrics.tolerations }}
+      tolerations:
+{{ toYaml .Values.metrics.tolerations | indent 8 }}
+      {{- end }}
+      containers:
+      - name: metrics
+        image: {{ template "metrics.image" . }}
+        imagePullPolicy: {{ .Values.metrics.image.pullPolicy | quote }}
+        env:
+        - name: DATA_SOURCE_NAME
```
Comment (tompizmor) [inline stable/postgresql/templates/metrics-deployment.yaml]: I think the metrics configuration is outdated. @javsalgar implemented this changes:

https://github.com/bitnami/charts/pull/875

**Review by tompizmor — CHANGES_REQUESTED** [review]

**Comment by carrodher** [comment]
Rebased applying last feature: https://github.com/helm/charts/commit/65403e85b0076ecbeffd2f4693cdca0ee59a0a46

**Review by tompizmor — APPROVED** [review]

**Review by javsalgar — APPROVED** [review]

**Comment by kfox1111** [comment]
Looks like bitnami's version also uses statefulset, which is a big selling point too.

File: stable/postgresql/.helmignore [hunk 23]
```diff
@@ -0,0 +1 @@
+.git
```
Comment (desaintmartin) [inline stable/postgresql/.helmignore:1]: Maybe mention OWNERS as well.

**Comment by Remigius2011** [comment]
Hi,

***idsclaimer: I just dropped here as I have noted that my approach of deploying postgresql via the non-Bitnami helm chart is broken.*** so please direct me to the location where I can report my problems if this isn't the right one.

I am deploying a small-scale postgresql database on a rancher 2.0 based k8s cluster using the helm chart (last version used: 0.19.0). I have discovered that two features I am using are missing in version 2.x:

* use of an existing secret (parameter `existingSecret`)
* use if an existing pvc (paramteter `persistence.existingClaim`)

furthermore, the pod's and service's name has changed to a include a prefix or suffix `postgres` (in my case it is now `postgres-postgres` and I haven't investigated whether it is a prefix or suffix yet) - this must be fixed by changing the host name in all locations where other resources access the database - but in a running cluster this might still be difficult to achieve and mandate downtime.

As I have also tought this to the participants of my [rancher workshop](https://github.com/Remigius2011/rancher-workshop), my students's workloads / future experiments are affected as well.

While I absolutely welcome the initiative to take properly care ot the postgresql helm chart, I think that it would be better to avoid disruption where possible - even if this may be expected by people familiar with the concept of semver.

Should I post a new issue or is there a chance at all that these concerns will be addressed? As of now, the only chance I see is sticking with v0.19 of the chart.

**Comment by desaintmartin** [comment]
I believe the need of using an existing claim is gone with the switch to a StatefulSet (you can create the PVC ahead of time corresponding to what the volumeClaimTemplate wants, and it will work).

I let other people answer for existing secret.

About incompatible changes, I absolutely agree with you, this is why I proposed to use another, new chart for this work (although I realize it would be misleading for most people). As for now, I'm going to use my own fork of this chart until I have time to upgrade all (production) projects with downtime. I suppose there is no other way than doing what is being done right now, since it follows a needed Kubernetes change (deployment -> statefulset) with unavoidable downtime.

Maybe the service names can be the sames than the old ones.

**Comment by Remigius2011** [comment]
@desaintmartin In my view, the introduction of a new postgresql image (bitnami instead of the 'official' one) with different features, base images, tagging and release cycle as well as a different set of k8s objects to be created would be reason enough to create another helm chart - even though this might confuse a few, it might be less disrupting for those who already use the old helm chart (like me). The more I'm looking at the new helm chart version, it looks to me like a completely different animal instead of an evolution from the old one. Nevertheless I fully agree that introducing a `StatefulSet` and support for clustered deployments (using streaming replication) is clearly an asset - but migration to it should imho be a conscious decision and not an accident (sorry for the clear wording I use here).

People who upgrade to a new version of the helm chart (which is in `rancher` - conveniently but dangerously - just a couple of mouse clicks away, and as it often works, people get used to it) will most likely end up with a failed deployment which they may or may not be able to rollback (I haven't tested this - I was so far lucky enough to deploy the new chart only on a freshly created cluster).

I don't know how a `StatefulSet` deals with a pre-existing pvc (the documentation says 'or pre-provisioned by an admin' but I dod not find out how to 'pre-provision' storage for a `StatefulSet`). If you think about upgrading an existing deployment, a full database migration (i.e. dump, transfer the dump and restore it) seems unavoidable to me. Maybe it would help if you could describe exactly how to create a pre-existing pvc and use it for deploying this helm chart.

Support for pre-existing secrets should imho be easy to implement - just take the existing one instead of creating a new one.

Maybe adding a clearly visible warning at the top of the description might help even though not everybody might look at it.

**Comment by cliedeman** [comment]
Hello,

I started work on a PR to add existing secret back #9532. Would be great if @carrodher could take a look.
The change is mostly trivial, the only problematic part is the notes file.

A section on how to migrate from pre version 2 would probably be a good idea.

Ciaran

**Comment by Remigius2011** [comment]
> A section on how to migrate from pre version 2 would probably be a good idea.

this and a description / facility to use pv / pvc storage (instead of storage classes).

**Comment by arianitu** [comment]
I have no idea how this was accepted. How can you claim stable/postgresql and suddenly change the underlying container entirely.

My own images were based off the official postgres image and no longer work with this chart because bitnami is entirely different.

I noticed replication was added, but it was added through a new container, meaning I can't even upgrade to this chart.

Helm is definitely off my list of things I would use in production. You guys should have made an entirely different chart, postgres-bitnami or something...

**Comment by carrodher** [comment]
> I have no idea how this was accepted. How can you claim stable/postgresql and suddenly change the underlying container entirely.
> 
> My own images were based off the official postgres image and no longer work with this chart because bitnami is entirely different.
> 
> I noticed replication was added, but it was added through a new container, meaning I can't even upgrade to this chart.
> 
> Helm is definitely off my list of things I would use in production. You guys should have made an entirely different chart, postgres-bitnami or something...

Hi, I just provide an answer in the issue that you have opened (https://github.com/helm/charts/issues/10451), to avoid duplicate conversations in different places (I have also seen a similar comment in another issue https://github.com/helm/charts/issues/8735#issuecomment-452122988), do you think it is good to continue the discussion in a single thread (https://github.com/helm/charts/issues/10451)?

**Comment by Remigius2011** [comment]
i second @arianitu 's comment about the disruption of the helm chart and also cannot understand how the (new) authors of the "official" postgresql helm chart could get away with changing it to be unusable with the official (this time without quotes) postgresql docker image by replacing it with a decommoditized one (namely the bitnami image, for which the official postgresql docker image is evidently not a drop-in replacement - even if this may be the case in the other direction). while using the bitnami image may undoubtedly have its benefits, it is imho unappropriate to force people to use it in place of the official one (which has also its benefits - e.g. early availability of new postgresql versions, comprehensible buildup etc..). what concerns me most is the disruption of operations caused by the upgrade of the helm chart from earlier versions. this has cost me - fortunately - only a few hours work, as i have first performed it on an experimental (not even test) environment, which was disposable, but this course of action has most likely cost some people who were performing the upgrade on a non-disposable environment severe headache and significant wasteful labor - which would have been absolutely avoidable.

while i think it would be appropriate to keep the "official" helm chart downwards compatible in the first place (which i myself tend to do / enforce / support in contexts where i have some control), the current situation - after a flood of new versions in short sequence - is probably already too much progressed in the new direction as to change the policy back to using the official postgresql docker image. however, it might be feasible to create a postgresql helm chart for the official docker image under a new name (maybe `postgresql-official`???) that forks off the last "usable" version (i am still at 0.19 and did not venture upgrading it to anything newer, as i have a productive environment running - i can still upgrade the docker image to newer official versions).

@arianitu , i don't think that helm itself is unusable, it has some benefits as it allows to create templates that bundle artifacts for complex deployments. this here is (imho) an accident that does not occur at large in the k8s/helm ecosystem (as far as i have observed). my lesson learned from it (which is not new, actually) is that upgrading helm charts lightheartedly should be avoided at any rate. instead it is advisable to perform experiments in a disposable environment first - which you **must** have at hand when you do anything serious anyway. this could happen in any situation where you change a component / artifact of your system, even when the evolution of components is performed as diligently as possible (this only lowers the likelyhood of disruption by some orders of magnitude instead of cuasing disruption cousciously).

@carrodher your question was not addressing me, but here's my two cents: i don't care where this discussion takes place (normally newer issues are designated as dupplicates for old ones, but there may be reasons to deviate from this practice), but my main concern is to keep up a (hopefully maintained) helm chart that supports the official postgresql docker image (i have currently no insight into the process of creating / maintaining "official" helm charts - probably it is documented just a few clicks away from here). obviously it would be advantageous for users of the old line of postgresql helm charts to have an upgrade path that avoids a complete reconstruction of the deployment (i.e. deploy new helm chart, migrate data and references from old to new etc.). if a new "official" image will not be made available, i will probably resort to forking off my own one as soon as the version i am currently using is not working any more (e.g. due to changes in k8s/helm).
